include/seqan3/range/decorator/gap_decorator.hpp

```diff
@@ -336,6 +336,152 @@ class gap_decorator
+        //!\brief Equals reference type.
+        using const_reference = reference;
+        //!\brief Pointer of container value type.
+        using pointer = value_type *;
+        //!\brief Signals that this is a random access iterator.
+        using iterator_category = std::random_access_iterator_tag;
+        //!\}
+
+        /*!\name Constructors/Destructors
+         * \{
+         */
+        constexpr gap_decorator_iterator_ra()                                              = default; //!< Defaulted.
+        constexpr gap_decorator_iterator_ra(gap_decorator_iterator_ra const &)             = default; //!< Defaulted.
+        constexpr gap_decorator_iterator_ra & operator=(gap_decorator_iterator_ra const &) = default; //!< Defaulted.
+        constexpr gap_decorator_iterator_ra (gap_decorator_iterator_ra &&)                 = default; //!< Defaulted.
+        constexpr gap_decorator_iterator_ra & operator=(gap_decorator_iterator_ra &&)      = default; //!< Defaulted.
+        ~gap_decorator_iterator_ra()                                                       = default; //!< Defaulted.
+
+        //!\brief Construct from seqan3::gap_decorator and initialise members.
+        explicit constexpr gap_decorator_iterator_ra(gap_decorator const & host_) noexcept : base_t{host_}
+        {}
+
+        //!\brief Construct from seqan3::gap_decorator and explicit position.
+        constexpr gap_decorator_iterator_ra(gap_decorator const & host_,
+                                            typename gap_decorator::size_type const pos_) noexcept : base_t{host_, pos_}
+        {}
+        //!\}
+
+        /*!\name Arithmetic operators
+        * \{
+        */
+        //!\brief Pre-increment, returns updated iterator.
+        constexpr gap_decorator_iterator_ra & operator++() noexcept
+        {
+            base_t::operator++;
+            return *this;
+        }
+
+        //!\brief Pre-decrement, returns updated iterator.
+        constexpr gap_decorator_iterator_ra & operator--() noexcept
+        {
+            base_t::operator--;
+            return *this;
+        }
+
+        //!\brief Post-increment, returns previous iterator state (delegates to pre-increment).
+        constexpr gap_decorator_iterator_ra operator++(int) noexcept
+        {
+            gap_decorator_iterator_ra cpy{*this};
+            ++(*this);
+            return cpy;
+        }
+
+        //!\brief Post-decrement, returns previous iterator state (delegates to pre-decrement).
+        constexpr gap_decorator_iterator_ra operator--(int) noexcept
+        {
+            gap_decorator_iterator_ra cpy{*this};
+            --(*this);
+            return cpy;
+        }
+
+        //!\brief Forward this iterator.
+        constexpr gap_decorator_iterator_ra & operator+=(difference_type const skip) noexcept
+        {
+            *this = gap_decorator_iterator_ra{*(this->host), this->pos + skip};
```

Comment:
Can you avoid reconstructing the whole object?
Like setting the new pos and trigger the gap counting?